_posts/2019-07-01-test-design.md

```diff
@@ -0,0 +1,294 @@
+
+## The most basic test design
+
+A common practice in writing tests is creating a test class for each production class they create. 
+The reason why it is such a popular practice, is because;
+* it's easy to find tests for the production code you're looking at,
+* it's a quick way to write new tests because you don't have to think about how and where to write tests.
+
+
+Although this approach does have advantages, it can also be harmful for the maintainability of your application.
+The disadvantage of this approach becomes clear when you need to refactor some classes.
+If you move logic from a class to another classes, or even multiple classes, you need to create new tests to test each of those classes, if you want to keep your 'one class means one test class' strategy.
+
+In performing such a refactoring we should not need to change any test because we are not adding or changing any functionality, only moving logic around. 
+However if we want to keep our *design* of having a test class for each production class, we need to refactor our tests as well.
+
+Even if we don't want to keep this design, our tests will have to be modified because chances are big that the api of our production code changed. 
+The parameters of methods might have changed, the fields of objects might have changed, constructors might have changed, etc.
+If we are lucky, the tests still compile, but they will very likely fail. 
+And the larger your application becomes, the more work it will be to get all tests compiling and green again.
+
+This phenomenon is known as [fragile tests](http://xunitpatterns.com/Fragile%20Test.html){:target="_blank" rel="noopener noreferrer"}.
+
+## Fragile tests example
+
+An example of this phenomenon that we encountered on a project is the creation of an instance of an [aggregate](https://martinfowler.com/bliki/DDD_Aggregate.html){:target="_blank" rel="noopener noreferrer"}.
+A lot of the tests in our project needed an instance of an aggregate. 
+This was not a problem at first, we just created aggregates by using the constructor of the class and passing all the necessary data in it.
+We created these instances in every test where we needed it, or sometimes created a method in the test class to not duplicate the construction too much in that class.
+
+To illustrate the issue we will look at a fictional simplified example about order creation.
+```java
+public class Order {
+    
+    private UUID customerId;
+    
+    public Order(final Customer customer) {
+        Assert.notNull(customer, "Customer should not be null");
+        Assert.isTrue(customer.isActive(), "Customer should be active");
+        this.customerId = customer.getId();
+    }
+    
+    public UUID getCustomerId() {
+        return this.customerId;
+    }
+}
+```
+
+```java
+public class OrderTest {
+    
+    @Test
+    public void createOrder_happyPath() {
```

Comment:
I personally like to see test names as sentences that describe what is being tested exactly.
While in Java we are stuck with the regular method naming restrictions we can utilise underscores to write our test name as a sentence.

```suggestion
    public void given_an_active_customer_then_order_creation_should_be_successful() {
```

In Kotlin for example this is slightly more convenient as you can write actual sentences by wrapping the method in backticks.